analyzer/evmverifier/sourcify/client.go

```diff
@@ -0,0 +1,210 @@
+	ethCommon "github.com/ethereum/go-ethereum/common"
+
+	"github.com/oasisprotocol/oasis-indexer/common"
+	"github.com/oasisprotocol/oasis-indexer/log"
+)
+
+const (
+	defaultServerUrl = "https://sourcify.dev/server"
+)
+
+// Oasis Runtime to Sourcify Chain ID mapping:
+// https://docs.sourcify.dev/docs/chains/
+var sourcifyChains = map[common.ChainName]map[common.Runtime]string{
+	common.ChainNameTestnet: {
+		common.RuntimeEmerald:  "42261",
+		common.RuntimeSapphire: "23295",
+	},
+	common.ChainNameMainnet: {
+		common.RuntimeEmerald:  "42262",
+		common.RuntimeSapphire: "23294",
+	},
+}
+
+// SourcifyClient is a client for interacting with the Sourcify Server API,
+// providing methods to fetch and parse contract data for supported Oasis runtimes.
+type SourcifyClient struct {
+	serverUrl *url.URL
+
+	chain common.ChainName
+
+	httpClient *http.Client
+
+	logger *log.Logger
+}
+
+func (s *SourcifyClient) callAPI(ctx context.Context, method string, url *url.URL) ([]byte, error) {
+	s.logger.Debug("sourcify API call", "method", method, "url", url.String())
+	req, err := http.NewRequestWithContext(ctx, method, url.String(), nil)
+	if err != nil {
+		return nil, err
+	}
+	res, err := s.httpClient.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("sourcify API call failure: %w", err)
+	}
+	defer res.Body.Close()
+	resp, err := io.ReadAll(res.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read response body: %w", err)
+	}
+
+	s.logger.Debug("sourcify API call response", "method", method, "url", url.String(), "status", res.Status, "response_body", string(resp))
+	if res.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("sourcify API call failure: %s", res.Status)
+	}
+
+	return resp, nil
+}
+
+// GetVerifiedContractAddresses returns a list of all verified contract addresses for the given runtime.
+//
+// API docs: https://docs.sourcify.dev/docs/api/server/get-contract-addresses-all/
+//
+// Note: This uses the free, public server API. If it turns out to be unreliable, we could use the repository API (vis IPFS proxy) instead, e.g.:
+// http://ipfs.default:8080/ipns/repo.sourcify.dev/contracts/full_match/23294
+func (s *SourcifyClient) GetVerifiedContractAddresses(ctx context.Context, runtime common.Runtime) ([]ethCommon.Address, error) {
+	// Fetch verified contract addresses.
+	u := *s.serverUrl
+	u.Path = path.Join(u.Path, "files/contracts", sourcifyChains[s.chain][runtime])
+	body, err := s.callAPI(ctx, http.MethodGet, &u)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch verified contract addresses: %w", err)
+	}
+
+	// Parse response.
+	var response struct {
+		Full []ethCommon.Address `json:"full"`
+		// XXX: Skip partial matches for now: https://docs.sourcify.dev/docs/full-vs-partial-match/
+	}
+	if err := json.Unmarshal(body, &response); err != nil {
+		return nil, fmt.Errorf("failed to parse verified contract addresses: %w", err)
+	}
+
+	return response.Full, nil
```

Comment:
Similar concern about scaling as with the DB. Maybe log a warning if this has more than 10k entries? Or if the API call takes more than 0.5s to return. Or something like that; something that allows us to revisit this in time.

Reply:
This only returns a list of addresses so I don't see it being problematic. 

Even for ethereum mainnet the response is "only" ~2.5MB in size (there's about ~50k addresses returned https://sourcify.dev/server/files/contracts/1/)